h2/src/main/org/h2/expression/ConditionAndOr.java

```diff
@@ -300,4 +307,40 @@ public Expression getSubexpression(int index) {
         }
     }
 
+    /**
+     * Optimize query according to the given condition. Example:
+     * (A AND B) OR (C AND B), the new condition (A OR C) AND B is returned
+     *
+     * @param left the session
+     * @param right the second condition
+     * @return null or the third condition
+     */
+    private Expression optimizeConditionAndOr(ConditionAndOr left, ConditionAndOr right){
+        if(left.andOrType == AND && right.andOrType == AND){
+            Expression commonExpression;
+            Expression combinedExpression;
+            if(left.getSubexpression(0).getSQL().equals(right.getSubexpression(0).getSQL())){
+                commonExpression = left.getSubexpression(0);
+                combinedExpression = new ConditionAndOr(OR,left.getSubexpression(1),right.getSubexpression(1));
+                return new ConditionAndOr(AND,commonExpression,combinedExpression);
+            }else if(left.getSubexpression(0).getSQL().equals(right.getSubexpression(1).getSQL())){
+                commonExpression = left.getSubexpression(0);
+                combinedExpression = new ConditionAndOr(OR,left.getSubexpression(1),right.getSubexpression(0));
+                return new ConditionAndOr(AND,commonExpression,combinedExpression);
+            }else if(left.getSubexpression(1).getSQL().equals(right.getSubexpression(0).getSQL())){
+                commonExpression = left.getSubexpression(1);
+                combinedExpression = new ConditionAndOr(OR,left.getSubexpression(0),right.getSubexpression(1));
+                return new ConditionAndOr(AND,commonExpression,combinedExpression);
+            }else if(left.getSubexpression(1).getSQL().equals(right.getSubexpression(1).getSQL())){
+                commonExpression = left.getSubexpression(1);
+                combinedExpression = new ConditionAndOr(OR,left.getSubexpression(0),right.getSubexpression(0));
+                return new ConditionAndOr(AND,commonExpression,combinedExpression);
+            }else {
+                return null;
+            }
+        }else {
+            return null;
+        }
+    }
+
 }
```

Comment:
please add a test-case to org.h2.test.db.TestOptimizations

Reply:
I will add the test case after a few moments. Thank you for the suggestions.

Reply:
Added a test case